### PR TITLE
Add support for vSphere volume mount/attach on Windows

### PR DESCRIPTION
**Pull request by gab-satchi**
**What type of PR is this?**
/kind feature

**What this PR does / why we need it**:
Adds support for attaching and mounting vSphere volumes on Windows. Analogous to this AWS PR https://github.com/kubernetes/kubernetes/pull/79552

**Special notes for your reviewer**:
ReadOnly volumes do not work. It seems like other Windows-compatible volume drivers also have this limitation right now.

**Does this PR introduce a user-facing change?**:
```release-note
Adds support for vSphere volumes on Windows
```

### CODE REVIEW

**Comment by PatrickLang** [comment]
LGTM, but I'm not a reviewer for vsphere

**Comment by michmike** [comment]
@saad-ali would you be able to review and approve this change?
thanks!

File: pkg/volume/vsphere_volume/attacher.go [hunk 1]
```diff
@@ -205,12 +206,17 @@ func (plugin *vsphereVolumePlugin) GetDeviceMountRefs(deviceMountPath string) ([
...
 	notMnt, err := mounter.IsLikelyNotMountPoint(deviceMountPath)
 	if err != nil {
 		if os.IsNotExist(err) {
-			if err := os.MkdirAll(deviceMountPath, 0750); err != nil {
-				klog.Errorf("Failed to create directory at %#v. err: %s", deviceMountPath, err)
+			dir := devicePath
+			if runtime.GOOS == "windows" {
+				dir = filepath.Dir(deviceMountPath)
+			}
+			if err := os.MkdirAll(dir, 0750); err != nil {
```
Comment (SandeepPissay) [inline pkg/volume/vsphere_volume/attacher.go]: For Linux OS - before this change, the code was creating directory on deviceMountPath. But now, we seem to be creating directory on devicePath. Could you explain the reason?

Comment (benmoss) [reply pkg/volume/vsphere_volume/attacher.go]: I think this might be an oversight, let me confirm

Comment (benmoss) [reply pkg/volume/vsphere_volume/attacher.go]: @SandeepPissay I confirmed that this was just a mistake and fixed the error

**Comment by SandeepPissay** [comment]
@divyenpatel Can you take a look at this pull request?

File: pkg/volume/vsphere_volume/vsphere_volume_util_unsupported.go [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package vsphere_volume
+
+import "errors"
+
+func verifyDevicePath(path string) (string, error) {
```
Comment (divyenpatel) [inline pkg/volume/vsphere_volume/vsphere_volume_util_unsupported.go]: I see build for this file  is `// +build !linux,!windows`, what are other supported build platform?

File: pkg/volume/vsphere_volume/attacher.go [hunk 3]
```diff
@@ -205,12 +206,17 @@ func (plugin *vsphereVolumePlugin) GetDeviceMountRefs(deviceMountPath string) ([
...
+	klog.Info("vsphere MountDevice", devicePath, deviceMountPath)
 	mounter := attacher.host.GetMounter(vsphereVolumePluginName)
 	notMnt, err := mounter.IsLikelyNotMountPoint(deviceMountPath)
 	if err != nil {
 		if os.IsNotExist(err) {
-			if err := os.MkdirAll(deviceMountPath, 0750); err != nil {
-				klog.Errorf("Failed to create directory at %#v. err: %s", deviceMountPath, err)
+			dir := deviceMountPath
+			if runtime.GOOS == "windows" {
+				dir = filepath.Dir(deviceMountPath)
```
Comment (divyenpatel) [inline pkg/volume/vsphere_volume/attacher.go]: later if error occurs for `diskMounter.FormatAndMount`
we do `os.Remove(deviceMountPath)`, may be for windows you need `os.Remove(filepath.Dir(deviceMountPath))`

File: pkg/volume/vsphere_volume/vsphere_volume.go [hunk 4]
```diff
@@ -239,9 +240,13 @@ func (b *vsphereVolumeMounter) SetUpAt(dir string, mounterArgs volume.MounterArg
...
 	}
 
-	if err := os.MkdirAll(dir, 0750); err != nil {
-		klog.V(4).Infof("Could not create directory %s: %v", dir, err)
-		return err
+	if runtime.GOOS != "windows" {
+		// On Windows, Mount will create the parent of dir and mklink (create a symbolic link) at dir later, so don't create a
+		// directory at dir now. Otherwise mklink will error: "Cannot create a file when that file already exists".
+		if err := os.MkdirAll(dir, 0750); err != nil {
+			klog.V(4).Infof("Could not create directory %s: %v", dir, err)
```
Comment (divyenpatel) [inline pkg/volume/vsphere_volume/vsphere_volume.go]: `klog.V(4).Infof` => `klog.Errorf`

**Review by divyenpatel — CHANGES_REQUESTED** [review]

File: pkg/volume/vsphere_volume/attacher.go (hunk 3 above)
Comment (benmoss) [reply pkg/volume/vsphere_volume/attacher.go]: There will be a symlink at `deviceMountPath`, I think we want to delete that, the same as on Linux. It looks like on both platforms the directory is left?

File: pkg/volume/vsphere_volume/vsphere_volume_util_unsupported.go (hunk 2 above)
Comment (benmoss) [reply pkg/volume/vsphere_volume/vsphere_volume_util_unsupported.go]: There aren't any other supported platforms, that's why this file is called `_unsupported.go`. This is a common pattern in Go when introducing platform-specific files, to include an `_unsupported` file that will still allow things to compile on others but leaves it up to someone who wants to make it work to add the necessary code.

**Review by divyenpatel — APPROVED** [review]

**Pull request merged**